### PR TITLE
Handle NPE when ITest implementation returns null

### LINKED ISSUE #993: NPE in JUnitXMLReporter
```
Failed to execute goal   org.apache.maven.plugins:maven-failsafe-plugin:2.19.1:integration-test (run-tests) on project XXXX: Execution run-tests of goal org.apache.maven.plugins:maven-failsafe-plugin:2.19.1:integration-test failed: There was an error in the forked process
java.lang.NullPointerException
at java.util.Hashtable.put(Hashtable.java:459)
at java.util.Properties.setProperty(Properties.java:166)
at org.testng.reporters.JUnitXMLReporter.createElement(JUnitXMLReporter.java:207)
at org.testng.reporters.JUnitXMLReporter.createElementFromTestResults(JUnitXMLReporter.java:187)
at org.testng.reporters.JUnitXMLReporter.generateReport(JUnitXMLReporter.java:178)
at org.testng.reporters.JUnitXMLReporter.onFinish(JUnitXMLReporter.java:118)
at org.testng.TestRunner.fireEvent(TestRunner.java:1239)
at org.testng.TestRunner.afterRun(TestRunner.java:1030)
at org.testng.TestRunner.run(TestRunner.java:636)
at org.testng.SuiteRunner.runTest(SuiteRunner.java:366)
at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:361)
at org.testng.SuiteRunner.privateRun(SuiteRunner.java:319)
at org.testng.SuiteRunner.run(SuiteRunner.java:268)
at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
at org.testng.TestNG.runSuitesSequentially(TestNG.java:1244)
at org.testng.TestNG.runSuitesLocally(TestNG.java:1169)
at org.testng.TestNG.run(TestNG.java:1064)
at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:132)
at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:99)
at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:147)
at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
 at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
 [Help 1]
```

From http://stackoverflow.com/q/35605443/4234729

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #993

Fixes #993 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/internal/TestResult.java [hunk 1]
```diff
@@ -101,26 +102,34 @@ private void init(ITestNGMethod method, ITestContext ctx, Throwable t, long star
     // toString() if it's been overridden.
     if (instance == null) {
       m_name = m_method.getMethodName();
+      return;
+    }
+    if (instance instanceof ITest) {
+      m_name = ((ITest) instance).getTestName();
+      if (m_name == null) {
```
Comment (juherr) [inline src/main/java/org/testng/internal/TestResult.java]: why not returning if `m_name != null`?

File: src/main/java/org/testng/internal/TestResult.java [hunk 2]
```diff
@@ -101,26 +102,34 @@ private void init(ITestNGMethod method, ITestContext ctx, Throwable t, long star
...
+      return;
+    }
+    if (instance instanceof ITest) {
+      m_name = ((ITest) instance).getTestName();
+      if (m_name == null) {
+        String msg = String.format(
+            "Warning: [%s] implementation on class [%s] returned null. Defaulting to method name",
+            ITest.class.getName(), instance.getClass().getName());
+        if (TestRunner.getVerbose() > 1) {
+          System.err.println(msg);
```
Comment (juherr) [inline src/main/java/org/testng/internal/TestResult.java]: You can use the logger class.

**Review by juherr — COMMENTED** [review]
Rework a bit the logic: quick return or failover to next strategy

File: src/main/java/org/testng/internal/TestResult.java (hunk 1 above)
Comment (krmahadevan) [reply src/main/java/org/testng/internal/TestResult.java]: fixed.

File: src/main/java/org/testng/internal/TestResult.java (hunk 2 above)
Comment (krmahadevan) [reply src/main/java/org/testng/internal/TestResult.java]: Using `Utils.log()` was my first choice. But I couldn't get the messages to show up on the console. That was why I chose the `syserr` route.

File: src/main/java/org/testng/internal/TestResult.java [hunk 3]
```diff
@@ -101,26 +102,35 @@ private void init(ITestNGMethod method, ITestContext ctx, Throwable t, long star
...
+      if (m_name != null) {
+        return;
+      }
+      String msg = String.format(
+          "Warning: [%s] implementation on class [%s] returned null. Defaulting to method name",
+          ITest.class.getName(), instance.getClass().getName());
+      if (TestRunner.getVerbose() > 1) {
+        System.err.println(msg);
+      }
+    } else if (method.getTestClass().getTestName() != null) {
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/internal/TestResult.java]: fixed.

**Review by juherr — APPROVED** [review]

**Pull request merged**